File: c++/Sort/InsertSort.cpp
#include<iostream>
using namespace std;

void InsertSort(int arr[] , int len){
    for( int j=0 ; j<len ; j++ ){
        int key = arr[j];//等待排序的第一个元素
        int i = j - 1; //代表已经排完序的元素的最后一个元素
        while( i>=0 && key<arr[i] ){
            // 从后向前比较已经排过序的数组乳沟比他小，则用后者代替前者。
            // 其实就是整个数组向后移动一位，以便找到合适的位置时便于插入
            arr[i+1] = arr[i];
            i--;
        }
        arr[i+1] = key; //找到合适的位置了，在I索引的后面设置KEY值
    }
}

int main(){
    int arr[10] = {1,2,3,4,5,22,3,7,1,2};
    for( int i=0 ; i<10 ; i++ ){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    
    InsertSort( arr , 10 );
    for( int i=0 ; i<10 ; i++ ){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    return 0;
}

File: go/Sort/InsertSort.go
package main

import "fmt"

func InsertSort(arr [10]int) [10]int {
	var i, j, tmp int
	for i = 1; i < len(arr); i++ {
		tmp = arr[i]
		for j = i; j > 0 && arr[j-1] > tmp; j-- {
			arr[j] = arr[j-1]
		}
		arr[j] = tmp
	}
	return arr
}

func main() {
	arr := [10]int{1, 4, 8, 7, 6, 5, 9, 10, 2, 3}
	fmt.Println("Source:", arr)

	arr = InsertSort(arr)

	fmt.Println("Sort:  ", arr)

}
